### PR TITLE
[BEAM-14470] Use lifecycle method names directly.

### PR DESCRIPTION
**Pull request by lostluck**
Use lifecycle method names directly, rather than excluding all exported methods on DoFns. 

Breaks certain usecases with extreme number (10+) of emitter parameters, when the method doesn't have a generic representation. The static code generator works in those cases, but it isn't recommended.

The ideal work around would be to have a tagged API instead of the positional one, but that's significantly more work.

Using lifecycle names directly neutered the previous verification check anyway, so we've opted to remove that entirely. Additional tests were also added to ensure the fallthrough behavior occurs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Review by jrmccluskey — APPROVED** [review]
LGTM, should get a look rom @damccorm as well since they should have more context

**Comment by lostluck** [comment]
As discussed offline, we decided to completely remove the method restriction entirely. While this may make things a little harder for novices, it will make things harder for novices with typos in the optional methods. These should be detectable by users via testing however.

Unfortunately there's no good way of validating this on our end short of doing some sort of typo/proximity algorithm, which is likely more trouble than better documentation and clearer examples, hence the initial strict approach.

**Comment by lostluck** [comment]
@damccorm PTAL

File: sdks/go/pkg/beam/core/graph/fn_test.go [hunk 1]
```diff
@@ -378,6 +380,129 @@ func TestNewCombineFn(t *testing.T) {
...
+	}
+	if dofn.TeardownFn() == nil {
+		t.Errorf("(%v).TeardownFn() == nil, want value", dofn.Name())
+	}
+	if dofn.IsSplittable() {
+		t.Errorf("(%v).IsSplittable() = true, want false", dofn.Name())
+	}
+}
+
+func TestNewFn_SplittableDoFn(t *testing.T) {
```
Comment (damccorm) [inline sdks/go/pkg/beam/core/graph/fn_test.go:427]: It would probably be good to add an extra similar test case for combineFns as well.

**Review by damccorm — COMMENTED** [review]
For posterity, we decided offline to remove `verifyValidNames` because the current change effectively circumvents that check in some cases and it doesn't make sense to only enforce it some of the time and/or leave behind an unenforced restriction. We're not planning on ever enforcing this restriction going forward (at least not without a major version bump).

**Pull request merged**